### PR TITLE
Extract JS code in FAQ doc and create testable faqlist service

### PR DESCRIPTION
**Pull request by mariavkourouni**
[FAQ doc test-link](http://mf-chsdi3.dev.bgdi.ch/kom_faqlist/api/faq/index.html)

[FAQ list JSON](http://mf-chsdi3.dev.bgdi.ch/kom_faqlist/rest/services/all/faqlist)

### CODE REVIEW

File: chsdi/views/faqlist.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# -*- coding: utf-8 -*-
+
+import requests
+from requests.exceptions import RequestException
+from pyramid.view import view_config
+from pyramid.httpexceptions import (HTTPNotFound, HTTPBadRequest)
+
+
+@view_config(route_name='faqlist', renderer='json')
```
Comment (loicgasser) [inline chsdi/views/faqlist.py]: we could use jsonp, it's useful sometimes

File: chsdi/views/faqlist.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+def faqlist(self):
+    try:
+        layersConfigURL = self.route_url('layersConfig', map='api')
```
Comment (loicgasser) [inline chsdi/views/faqlist.py]: please use db directly and avoid round trips.

```python
from chsdi.models import LayersConfig
query = request.db.query(LayersConfig)
```

File: chsdi/views/faqlist.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                if 'chargeable' in layer and not layer['chargeable']:
+                    try:
+                        no_chargeable_layers.append(layer_name)
+                    except:
+                        pass
+        except RequestException:
+            pass
+
+    if (len(queryable_layers) + len(searchable_layers) + len(chargeable_layers) + len(no_chargeable_layers)) == 0:
+        raise HTTPBadRequest()
```
Comment (loicgasser) [inline chsdi/views/faqlist.py]: why is that? `HTTPBadRequest` should be not thrown here. You can use that when parameters are bad/wrong.

File: chsdi/views/faqlist.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    chargeable_layers = []
+    no_chargeable_layers = []
+
+    for layer_name in layers_api:
+        try:
+            layer = layers_api[layer_name]
+            if 'parentLayerId' not in layer and not str(layer_name).endswith(suffix3d):
+                if 'queryable' in layer and layer['queryable']:
+                    try:
+                        queryable_layers.append(layer_name)
```
Comment (loicgasser) [inline chsdi/views/faqlist.py]: what are you afraid of here? no need to `try/except` every time.

**Comment by loicgasser** [comment]
I'll come back to it later when adapted. come to my office if you have questions.

**Comment by loicgasser** [comment]
@mariavkourouni I was thinking that you can also adapt the faq page at the same time.

File: chsdi/views/layers.py [hunk 5]
```diff
@@ -160,6 +160,56 @@ def insertValueAt(field, attrName, value):
...
+                notChargeableLayers.append(l)
+
+    queryableLayers.sort()
+    tooltipLayers.sort()
+    searchableLayers.sort()
+    chargeableLayers.sort()
+    notChargeableLayers.sort()
+    return {
+        'translations': translations,
+        'queryableLayers': queryableLayers,
```
Comment (loicgasser) [inline chsdi/views/layers.py]: for less code simply do:
```python
queryableLayers': queryableLayers.sort(),
```

Comment (mariavkourouni) [inline chsdi/views/layers.py]: Like this it was returning null, probably because of the comma exactly after `.sort()`

**Comment by loicgasser** [comment]
yes makes sense that it returns null, good alternative.

Looking good. We just need 1 or 2 integration test for this service and I ll merge. thanks! @gjn maybe could have a look to make sure this is what you want.

**Comment by AFoletti** [comment]
What's the better way to test it?

**Comment by loicgasser** [comment]
@AFoletti 

2 ways to do that:
1) Test that the service is responding correctly with integration tests
2) When deploying to int create a diff with the previous version of this service to check that the new/updated layers possess the correct features.

**Comment by gjn** [comment]
Both 1) and 2) is part of @mariavkourouni tasks.

**Comment by mariavkourouni** [comment]
yes, I am going to do them!

**Comment by mariavkourouni** [comment]
1) Integration tests are done for the service
2) Diff-script is ready, so when deploying to int, the comparison between phases could be implemented

**Comment by loicgasser** [comment]
@mariavkourouni looks good please squash and rebase

File: chsdi/tests/integration/test_mapservice.py [hunk 6]
```diff
@@ -34,6 +34,14 @@ def test_metadata_chargeable_false(self):
...
         self.assertEqual(resp.content_type, 'application/json')
 
+    def test_faqlist(self):
+        resp = self.testapp.get('/rest/services/api/faqlist', status=200)
+        self.assertEqual(resp.content_type, 'application/json')
+
+    def test_faqlist_topic_all(self):
+        resp = self.testapp.get('/rest/services/all/faqlist', status=200)
+        self.assertEqual(resp.content_type, 'application/json')
+
```
Comment (gjn) [inline chsdi/tests/integration/test_mapservice.py]: Please check here that we have at least x layers in the results. (x high, but 2-3 smaller than current count).

Also, check for correct structure (by checking for the 4 lists, and 1-2 layers in each list)

File: chsdi/static/doc/source/api/faq/index.rst [hunk 7]
```diff
@@ -173,16 +123,16 @@ Not all the layers have a tooltip. Here is a complete list of all the layers tha
 Which layers are searchable?
 ****************************
 
-We define a layer as searchable when its features can be searched. Here is a list of all searchable layers:
+We define a layer as searchable when its features can be searched. Here there is the list of all searchable layers:
```
Comment (loicgasser) [inline chsdi/static/doc/source/api/faq/index.rst]: you corrected the language but I think it's wrong

File: chsdi/static/doc/source/api/faq/index.rst [hunk 8]
```diff
@@ -173,16 +123,16 @@ Not all the layers have a tooltip. Here is a complete list of all the layers tha
...
 .. raw:: html
 
   <body>
-    <div id="searchable" style="margin-left:10px;margin-bottom:24px;"></div>
+    <div id="searchableLayers" style="margin-left:10px;margin-bottom:24px;"></div>
   </body>
 
-How can I accessed the tiles ?
-******************************
+How can I have access to the tiles ?
```
Comment (loicgasser) [inline chsdi/static/doc/source/api/faq/index.rst]: :+1:

File: chsdi/static/doc/source/api/faq/index.rst (hunk 7 above)
Comment (loicgasser) [inline chsdi/static/doc/source/api/faq/index.rst]: ok I fix myself

**Comment by gjn** [comment]
Thanks alot @mariavkourouni and @loicgasser . Good improvement.

**Pull request merged**